ccl/changefeedccl: TestChangefeedHandlesDrainingNodes failed

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9482889?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9482889?buildTab=artifacts#/) on release-23.1 @ [571a1744c72415894dc0e742bd89e74938ea71f4](https://github.com/cockroachdb/cockroach/commits/571a1744c72415894dc0e742bd89e74938ea71f4):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/905465f697dc04c39f7d8a82cb057bfe/logTestChangefeedHandlesDrainingNodes3712854043
    test_log_scope.go:79: use -show-logs to present logs inline
    helpers_test.go:895: making webhook feed factory
    changefeed_test.go:6055: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:195
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:270
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_test.go:6055
        	Error:      	Received unexpected error:
        	            	Details for job 854521624819171329: pq: system-jobs-scan: could not register flowID 4a24d537-00fc-4e4e-9115-0ac45b500bf2 because the registry is draining
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*jobFeed).Details
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/testfeed_test.go:468
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*webhookFeed).Next
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/testfeed_test.go:2126
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.readNextMessages
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:101
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBaseErr
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:206
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:198
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:263
        	            	  | github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:91
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:259
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:196
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloads
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:270
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestChangefeedHandlesDrainingNodes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_test.go:6055
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) Details for job 854521624819171329
        	            	Wraps: (3) pq: system-jobs-scan: could not register flowID 4a24d537-00fc-4e4e-9115-0ac45b500bf2 because the registry is draining
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *pq.Error
        	Test:       	TestChangefeedHandlesDrainingNodes
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/905465f697dc04c39f7d8a82cb057bfe/logTestChangefeedHandlesDrainingNodes3712854043
--- FAIL: TestChangefeedHandlesDrainingNodes (8.12s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-26682

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9506718?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9506718?buildTab=artifacts#/) on release-23.1 @ [de239a7438f44d382c9aefceb65d9c39911dabd2](https://github.com/cockroachdb/cockroach/commits/de239a7438f44d382c9aefceb65d9c39911dabd2):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1c3ce425109520152572148b124af7f1/logTestChangefeedHandlesDrainingNodes2127889315
    test_log_scope.go:79: use -show-logs to present logs inline
    helpers_test.go:895: making pubsub feed factory
    changefeed_test.go:6056: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:195
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:270
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_test.go:6056
        	Error:      	Received unexpected error:
        	            	Details for job 854811279810789377: pq: system-jobs-scan: could not register flowID c5a93cf8-208c-4492-b506-713bcd61b350 because the registry is draining
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*jobFeed).Details
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/testfeed_test.go:473
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*pubsubFeed).Next
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/testfeed_test.go:2489
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.readNextMessages
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:101
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBaseErr
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:206
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:198
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:263
        	            	  | github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:91
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:259
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:196
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloads
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:270
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestChangefeedHandlesDrainingNodes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_test.go:6056
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) Details for job 854811279810789377
        	            	Wraps: (3) pq: system-jobs-scan: could not register flowID c5a93cf8-208c-4492-b506-713bcd61b350 because the registry is draining
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *pq.Error
        	Test:       	TestChangefeedHandlesDrainingNodes
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1c3ce425109520152572148b124af7f1/logTestChangefeedHandlesDrainingNodes2127889315
--- FAIL: TestChangefeedHandlesDrainingNodes (6.23s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9515388?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9515388?buildTab=artifacts#/) on release-23.1 @ [cfbd751ae4c82e2c4b5314376cd6e6c29cc30aea](https://github.com/cockroachdb/cockroach/commits/cfbd751ae4c82e2c4b5314376cd6e6c29cc30aea):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/8664cbf2d4e437e0e06ce871fe9210c9/logTestChangefeedHandlesDrainingNodes1936342772
    test_log_scope.go:79: use -show-logs to present logs inline
    helpers_test.go:895: making pubsub feed factory
    changefeed_test.go:6048: failed to start feed for job 0: pq: opening client: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8664cbf2d4e437e0e06ce871fe9210c9/logTestChangefeedHandlesDrainingNodes1936342772
--- FAIL: TestChangefeedHandlesDrainingNodes (8.21s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9524014?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9524014?buildTab=artifacts#/) on release-23.1 @ [2f96695f75b07c872ec5f146acc1fa198135768f](https://github.com/cockroachdb/cockroach/commits/2f96695f75b07c872ec5f146acc1fa198135768f):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/8664cbf2d4e437e0e06ce871fe9210c9/logTestChangefeedHandlesDrainingNodes144194294
    test_log_scope.go:79: use -show-logs to present logs inline
    helpers_test.go:895: making pubsub feed factory
    changefeed_test.go:6048: failed to start feed for job 0: pq: opening client: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8664cbf2d4e437e0e06ce871fe9210c9/logTestChangefeedHandlesDrainingNodes144194294
--- FAIL: TestChangefeedHandlesDrainingNodes (11.75s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/changefeedccl.TestChangefeedHandlesDrainingNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9525422?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9525422?buildTab=artifacts#/) on release-23.1 @ [416beb787ce5cf90695f13ebb7c95f749dc14061](https://github.com/cockroachdb/cockroach/commits/416beb787ce5cf90695f13ebb7c95f749dc14061):


```
=== RUN   TestChangefeedHandlesDrainingNodes
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/8664cbf2d4e437e0e06ce871fe9210c9/logTestChangefeedHandlesDrainingNodes701142468
    test_log_scope.go:79: use -show-logs to present logs inline
    helpers_test.go:895: making pubsub feed factory
    changefeed_test.go:6048: failed to start feed for job 0: pq: opening client: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8664cbf2d4e437e0e06ce871fe9210c9/logTestChangefeedHandlesDrainingNodes701142468
--- FAIL: TestChangefeedHandlesDrainingNodes (9.32s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedHandlesDrainingNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -5972,7 +5972,7 @@ func TestChangefeedHandlesDrainingNodes(t *testing.T) {
 	sqlDB.Exec(t, "ALTER TABLE test.foo SCATTER")
 
 	// Create a factory which executes the CREATE CHANGEFEED statement on server 0.
-	// This statement should fail, but the job itself ought to be creaated.
+	// This statement should fail, but the job itself ought to be created.
 	// After some time, that job should be adopted by another node, and executed successfully.
 	f, closeSink := makeFeedFactory(t, randomSinkType(feedTestEnterpriseSinks), tc.Server(1), tc.ServerConn(0))
 	defer closeSink()
@@ -5981,22 +5981,21 @@ func TestChangefeedHandlesDrainingNodes(t *testing.T) {
 	feed := feed(t, f, "CREATE CHANGEFEED FOR foo")
 	defer closeFeed(t, feed)
 
-	// At this point, the job created by feed will fail to start running on node 0 due to draining
-	// registry.  However, this job will be retried, and it should succeed.
-	// Note: This test is a bit unrealistic in that if the registry is draining, that
-	// means that the server is draining (i.e. being shut down).  We don't do a full shutdown
-	// here, but we are simulating a restart by failing to start a flow the first time around.
-	assertPayloads(t, feed, []string{
-		`foo: [1]->{"after": {"k": 1, "v": 1}}`,
-		`foo: [2]->{"after": {"k": 2, "v": 0}}`,
-		`foo: [3]->{"after": {"k": 3, "v": 1}}`,
-		`foo: [4]->{"after": {"k": 4, "v": 0}}`,
-		`foo: [5]->{"after": {"k": 5, "v": 1}}`,
-		`foo: [6]->{"after": {"k": 6, "v": 0}}`,
-		`foo: [7]->{"after": {"k": 7, "v": 1}}`,
-		`foo: [8]->{"after": {"k": 8, "v": 0}}`,
-		`foo: [9]->{"after": {"k": 9, "v": 1}}`,
-		`foo: [10]->{"after": {"k": 10, "v": 0}}`,
+	jobID := feed.(cdctest.EnterpriseTestFeed).JobID()
+	registry := tc.Server(1).JobRegistry().(*jobs.Registry)
+	loadProgress := func() jobspb.Progress {
+		job, err := registry.LoadJob(context.Background(), jobID)
+		require.NoError(t, err)
+		return job.Progress()
+	}
+
+	// Wait until highwater advances.
+	testutils.SucceedsSoon(t, func() error {
+		progress := loadProgress()
+		if hw := progress.GetHighWater(); hw == nil || hw.IsEmpty() {
+			return errors.New("waiting for highwater")
+		}
+		return nil
 	})
 }
 

